File: Python/Copy List with Random Pointer.py
# A linked list is given such that each node contains an additional random pointer which could point to any node in the list or null.

# Return a deep copy of the list.


# Definition for singly-linked list with a random pointer.
# class RandomListNode:
#     def __init__(self, x):
#         self.label = x
#         self.next = None
#         self.random = None

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):
		current = head
		while current != None:
			temp = RandomListNode(current.label)
			temp.next = current.next
			current.next = temp
			current = temp.next
		current = head
		while current != None:
			temp = current.next
			if current.random != None:
				temp.random = current.random.next
			current = temp.next
		current = head
		if head == None: 
			dup = None 
		else: 
			dup = head.next
		while current != None:
			temp = current.next
			current.next = temp.next
			if temp.next != None:
				temp.next = current.next.next
			current = current.next
		return dup

class Solution:
    # @param head, a RandomListNode
    # @return a RandomListNode
    def copyRandomList(self, head):

File: Java/Copy_List_with_Random_Pointer.java
class Copy_List_with_Random_Pointer{
    public RandomListNode copyRandomList(RandomListNode head) {
        if(head == null)
        	return null;
        RandomListNode p = head;
        while(p != null){
        	RandomListNode node = new RandomListNode(p.label);
        	node.next = p.next;
        	p.next = node;
        	p = node.next;
        }
        p = head;
        while(p != null){
        	if(p.random != null){
        		p.next.random = p.random.next;
        	}
        	else{
        		p.next.random = null;
        	}
        	p = p.next.next;
        }
        RandomListNode newHead = head.next;
        p = head;
        while(p != null){
        	RandomListNode dup = p.next;
        	p.next = dup.next;
        	if(dup.next != null)
        		dup.next = dup.next.next;
        	else
        		dup.next = null;
        	p = p.next;
        }
        return newHead;
    }	
}